### PR TITLE
[FLINK-23680][streaming] Wait until all expected threads would be triggered in testOpenCloseAndTimestamps

### PR DESCRIPTION
**Pull request by akalash**
## What is the purpose of the change

*This PR fixes the test StreamTaskTimerTest.testOpenCloseAndTimestamps*

## Brief change log

  - *Waiting for all threads which should be triggered during the task start before spawning the new thread for the test*

## Verifying this change

It is test itself

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by curcur** [comment]
Hey @akalash , the title of this PR is broken into two lines. 

The test itself is strange. I am not sure whether the second schedule will definitely spawn a new thread, so maybe let's remove it and only use the one in the invoke?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java [hunk 1]
```diff
@@ -70,11 +70,22 @@ public void teardown() throws Exception {
...
 
-        assertEquals(2, StreamTask.TRIGGER_THREAD_GROUP.activeCount());
+        assertEquals(initiallyExpectedThreads + 1, StreamTask.TRIGGER_THREAD_GROUP.activeCount());
+    }
+
+    private void waitUntilTaskTriggerThreadsOnStart(int expectedThreads)
+            throws InterruptedException {
+        while (StreamTask.TRIGGER_THREAD_GROUP.activeCount() != expectedThreads) {
+            Thread.sleep(1);
+        }
```
Comment (curcur) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java:88]: Add a timeout here?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java [hunk 2]
```diff
@@ -70,11 +70,22 @@ public void teardown() throws Exception {
...
     @Test
-    public void testOpenCloseAndTimestamps() {
-        // first one spawns thread
+    public void testOpenCloseAndTimestamps() throws InterruptedException {
+        // Wait for the start of threads: throughput calculation.
+        int initiallyExpectedThreads = 1;
+        waitUntilTaskTriggerThreadsOnStart(initiallyExpectedThreads);
+
+        // Spawn one more thread.
         timeService.registerTimer(System.currentTimeMillis(), timestamp -> {});
```
Comment (curcur) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java:79]: Maybe removing this line directly, and use the scheduled calculating throughput directly?

I am not sure scheudle a new timer will 100% to spawn a new thread.

**Review by curcur — CHANGES_REQUESTED** [review]

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java (hunk 1 above)
Comment (akalash) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java]: I added the timeout for test

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java (hunk 2 above)
Comment (akalash) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java]: I rewrote the test

**Review by curcur — APPROVED** [review]

**Comment by curcur** [comment]
Thanks for fixing this!

LGTM. The failure does not seem to relate to this fix.

**Pull request merged**